lib/poecfg.cpp

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2024
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "poecfg.h"
+#include "tokenize.h"
+
+namespace swss {
+
+std::string PoeConfig::parseKey(const std::string &key_str)
+{
+    std::vector<std::string> token = tokenize(key_str, ':');
+    std::string type, idx;
+
+    if (token.size() == 1)
+    {
+        if (token[0].rfind("Ethernet", 0) != 0)
+        {
+            return "";
+        }
+        type = "port";
+        idx = token.at(0);
+    }
+    else if (token.size() == 2)
+    {
+        type = token.at(0);
+        idx = token.at(1);
+    }
+
+    SWSS_LOG_NOTICE("Parsed key:%s, Return:(%s)(%s)", key_str.c_str(), type.c_str(), idx.c_str());
+
+    return type;
+}
+
+PoeConfig::PoeConfig(Table *poe_table)
+{
+    poeTable = poe_table;
+}
+
+PoeConfig::PoeConfig(Table &poe_table)
+{
+    poeTable = &poe_table;
+}
+
+bool PoeConfig::isConfigDone()
+{
+    std::vector<FieldValueTuple> tuples;
+
+    if (!poeTable->get("PoeConfigDone", tuples))
+    {
+        return false;
+    }
+
+    for (auto& el : tuples)
+    {
+        if (el.first == "success" && el.second == "true")
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool PoeConfig::platformHasPoe()
+{
+    bool ret = false;
+
+    if (access("/usr/share/sonic/hwsku/poe_config.json", F_OK) != -1)
+    {
+        ret = true;
+    }
+    return ret;
+}
+
+bool PoeConfig::isPoeEnabled()
+{
+    int count_s = 10;
+
+    SWSS_LOG_ENTER();
+    if (!platformHasPoe())
+    {
+        return false;
+    }
+
+    while (count_s > 0)  /* wait for 10s */
+    {
+        if (isConfigDone())
+        {
+            return true;
+        }
+        sleep(1);
+        count_s--;
+        continue;
+    }
+
+    return false;
+}
+
+bool PoeConfig::loadDeviceConfig(const std::vector<FieldValueTuple> &ovalues)
```

Comment:
The loadDeviceConfig, loadPseConfig, and loadPortConfig functions always return true, even though some fields might be missing or invalid.
Could you add error handling that logs and returns false?

Reply:
added exception handling (for std::stoul()) and error logs